### PR TITLE
VoxelArea: faster iter function

### PR DESCRIPTION
**Pull request by HybridDog**
a big cuboid(
		local x2 = x1+97
		local y2 = y1+40
		local z2 = z1+30
)
worked ~4.3 times as fast as before
and a small one(
		local x2 = x1+2
		local y2 = y1+4
		local z2 = z1+3
) worked ~1.2 times as fast as before

### CODE REVIEW

**Comment by asl97** [comment]
I wrote a little standalone benchmark, hopefully it's correct.
The big cuboid was about 3 times faster on lua and 8 times faster on luajit(?)*.
____

I tried implemented it using coroutine thinking it could run better but it turns out worse than others xD
I included it in the test just for fun.

Github require text file to be named with '.txt': [test.txt](https://github.com/minetest/minetest/files/445789/test.txt)

```
$ lua test.txt

cube: x = 3, y = 5, z = 4
volume:	60

5e-05 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
3.4e-05 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
7.1e-05 : asl97 coroutine iter


cube: x = 11, y = 101, z = 11
volume:	12221

0.006312 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
0.002876 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
0.004926 : asl97 coroutine iter


cube: x = 98, y = 41, z = 31
volume:	124558

0.042927 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
0.016867 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
0.042846 : asl97 coroutine iter

$ luajit test.txt

cube: x = 3, y = 5, z = 4
volume:	60

0.000127 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
0.000106 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
0.000208 : asl97 coroutine iter


cube: x = 11, y = 101, z = 11
volume:	12221

0.001229 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
0.001093 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
0.002769 : asl97 coroutine iter


cube: x = 98, y = 41, z = 31
volume:	124558

0.008031 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
0.005029 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
0.013244 : asl97 coroutine iter
```

*Note: Running the last test by itself (by commenting out the first two) on luajit is a lot faster for some reason.
```
$ luajit test.txt

cube: x = 98, y = 41, z = 31
volume:	124558

0.002754 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
0.000343 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
0.013832 : asl97 coroutine iter

```

**Comment by est31** [comment]
Nice, I like it.

**Comment by HybridDog** [comment]
asl97, benchmarking is much more reliable (correct me if l'm wrong) if you count how often it executes in TIME seconds instead of measuring how long it takes if it's executed one time
http://dev.minetest.net/Lua_Optimization_Tips#Benchmarking

**Comment by asl97** [comment]
@HybridDog That is true, but normally one would run it a number of time and the best (average) time. instead of running just once (kind of like python timeit module).

I believe ether way is fine, some prefer how much time it takes to execute x times while other prefer how many times it can execute in x seconds.

it's just the formal is easier to implement correctly.
____
benchmark 2.0,  number of executes per second: [test2.txt](https://github.com/minetest/minetest/files/447051/test2.txt)
```
$ lua test2.txt

cube: x = 3, y = 5, z = 4
volume:	60

43251.0 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
65466.803599589 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
36837.189581829 : asl97 coroutine iter


cube: x = 11, y = 101, z = 11
volume:	12221

255.15111224955 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
550.89532988732 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
237.06808535646 : asl97 coroutine iter


cube: x = 98, y = 41, z = 31
volume:	124558

24.932532566874 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
59.420648675415 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
24.145304510922 : asl97 coroutine iter

$ luajit test2.txt

cube: x = 3, y = 5, z = 4
volume:	60

306225.69377431 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
411341.58865841 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
95460.618157527 : asl97 coroutine iter


cube: x = 11, y = 101, z = 11
volume:	12221

3776.2372000856 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
16406.688272923 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
686.12381988201 : asl97 coroutine iter


cube: x = 98, y = 41, z = 31
volume:	124558

377.15592503976 : iter from 92815ad54b23fe92742ebca7263bb227149248c1
3342.0942924467 : iter from HybridDog:f0d248e45ce64c3b99d3c03524ebb9ee3bed7790
68.623258311868 : asl97 coroutine iter
```

**Comment by HybridDog** [comment]
The iter function could become further speeded up by testing if x, y or z ranges are 1,
but it would slow down (more or less slightly) the function (not the returned one) if no range is 1.

**Comment by paramat** [comment]
Nice method. I assume you tested that it creates a complete solid cuboid? If so +1.

**Comment by HybridDog** [comment]
>I assume you tested that it creates a complete solid cuboid?

Yes, l did, the function is btw derivated from #4479.

**Comment by paly2** [comment]
LGTM :+1:

**Comment by paramat** [comment]
:+1:

**Comment by paramat** [comment]
@est31 is that an approval?

**Comment by paramat** [comment]
Hang on, some code style issues ..

File: builtin/game/voxelarea.lua [hunk 1]
```diff
@@ -84,23 +84,46 @@ end
...
+	local multistride = self.zstride - ((yrange - 1) * self.ystride + xrange)
+
 	return function()
-		i = i + 1
-		if i % zstride == zoff then
-			i = i + zstridediff
-		elseif i % ystride == yoff then
-			i = i + ystridediff
+		-- continue i until it needs to jump
+		i = i+1
```
Comment (paramat) [inline builtin/game/voxelarea.lua]: spaces

File: builtin/game/voxelarea.lua [hunk 2]
```diff
@@ -84,23 +84,46 @@ end
...
+		if i ~= nextaction then
+			return i
 		end
-		if i <= last then
+
+		-- continue y until maxy is exceeded
+		y = y+1
```
Comment (paramat) [inline builtin/game/voxelarea.lua]: spaces

File: builtin/game/voxelarea.lua [hunk 3]
```diff
@@ -84,23 +84,46 @@ end
...
+		if y ~= yrange then
+			-- set i to index(minx, miny + y, minz + z) - 1
+			i = i + yreqstride
+			nextaction = i + xrange
 			return i
 		end
+
+		-- continue z until maxz is exceeded
+		z = z+1
```
Comment (paramat) [inline builtin/game/voxelarea.lua]: spaces

**Comment by paramat** [comment]
Minor code issues, otherwise +1.

**Comment by est31** [comment]
@paramat no, no approval. Just stated that I liked it nothing more :)

**Comment by paramat** [comment]
:+1:

**Pull request merged**